source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -334,9 +315,7 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
     // dedicated thread in a separate API within which handshake_complete_
     // will be set to true if the handshake completes.
     ASSERT(!handshake_complete_);
-    ASSERT(action == Network::PostIoAction::KeepOpen);
-    // TODO(lizan): Handle synchronous handshake when TsiHandshaker supports it.
-    return {Network::PostIoAction::KeepOpen, 0, false};
+    return {action, 0, action == Network::PostIoAction::Close ? true : false};
```

Comment:
besides, is there an integration test cover this?

Reply:
Fixed.

AFAIK, there is no integration test in OSS Envoy that caught this. If you can point me to an example to base of, I'm happy to add one (or maybe we should defer this to a follow-up PR since this is an existing issue).

Reply:
I am thinking like we should trigger the assertion if this line is triggered. We can leverage https://github.com/envoyproxy/envoy/blob/47afc8c884ca95a40d01513cf82f0f40b67b8ed8/test/extensions/transport_sockets/alts/alts_integration_test.cc#L195
to run into this condition during `doWrite` call.

Reply:
Sorry I missed this. There are tests in tsi_socket_test.cc that check this code block now.

Reply:
Thanks!